File: Set 7/Minimum Sum Path/MSP.py
def helper(i, j, matrix, dp):
    if i == 0 and j == 0:
        return matrix[0][0]
    if i < 0 or j < 0:
        return int(1e9)
    if dp[i][j] != -1:
        return dp[i][j]
    up = matrix[i][j] + helper(i-1, j, matrix, dp)
    left = matrix[i][j] + helper(i, j-1, matrix, dp)
    dp[i][j] = min(up, left)
    return dp[i][j]

def minSumPath(n, m, matrix):
    dp = [[-1 for j in range(m)] for i in range(n)]
    return helper(n-1, m-1, matrix, dp)



t = int(input())
for _ in range(t):
    n = int(input("Enter the number of rows in the matrix: "))
    m = int(input("Enter the number of columns in the matrix: "))
    matrix = []
    print("Enter the matrix row by row:")
    for i in range(n):
        row = list(map(int, input().split()))
        matrix.append(row)
    print("The minimum sum path is:", minSumPath(n, m, matrix))

File: Set 7/Minimum Sum Path/MSP.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

int min(int x, int y, int z) {
    if (x < y)
        return (x < z) ? x : z;
    else
        return (y < z) ? y : z;
}

int minCost(vector<vector<int>>& cost, int m, int n) {
    if (m < 0 || n < 0)
        return INT_MAX;
    else if (m == 0 && n == 0)
        return cost[m][n];
    else
        return cost[m][n] + min(minCost(cost, m - 1, n - 1),
                                minCost(cost, m - 1, n),
                                minCost(cost, m, n - 1));
}

int main() {
    int T;
    cout << "Enter the number of test cases: ";
    cin >> T;

    while (T--) {
        int R, C;
        cout << "Enter the number of rows and columns: ";
        cin >> R >> C;

        vector<vector<int>> cost(R, vector<int>(C));
        cout << "Enter the cost matrix row by row:\n";
        for (int i = 0; i < R; ++i) {
            for (int j = 0; j < C; ++j) {
                cin >> cost[i][j];
            }
        }

        int m, n;
        cout << "Enter the target cell indices (m, n): ";
        cin >> m >> n;

        cout << "Minimum cost to reach (" << m << ", " << n << "): " << minCost(cost, m, n) << endl;
    }

    return 0;
}
